src/dbnode/persist/fs/retriever.go

```diff
@@ -376,27 +376,17 @@ func (r *blockRetriever) fetchBatch(
 		// then errSeekNotCompleted is returned because req.success is not set
 		// rather than we have dangling goroutines stacking up.
 		for _, req := range allReqs {
-			if req.waitOnFetch == nil {
+			if !req.waitingForCallback {
 				req.onDone()
 				continue
 			}
 
 			filteredReqs = append(filteredReqs, req)
 		}
 
-		allReqs = filteredReqs
-		for len(allReqs) > 0 {
-			filteredReqs = filteredReqs[:0]
-			for _, req := range allReqs {
-				select {
-				case <-req.waitOnFetch:
-					req.onDone()
-				default:
-					filteredReqs = append(filteredReqs, req)
-				}
-			}
-
-			allReqs = filteredReqs
+		for _, req := range filteredReqs {
+			<-req.waitForCallback
```

Comment:
Now that we are effectively waiting for the slowest `req` to complete, wouldn't it be simpler to implement this by waiting on a `WaitGroup` (and call `onDone()` for all the `req`s after it), and drop all those channels, and `waitingForCallback` flags?